File: src/main/java/cspfj/generator/constraint/AllDifferentGenerator.java
package cspfj.generator.constraint;

import java.util.List;

import com.google.common.base.Preconditions;
import com.google.common.collect.Iterables;
import com.google.common.collect.Lists;

import cspfj.constraint.semantic.AllDifferent;
import cspfj.exception.FailedGenerationException;
import cspfj.problem.Problem;
import cspfj.problem.Variable;
import cspom.constraint.CSPOMConstraint;
import cspom.constraint.GeneralConstraint;

public final class AllDifferentGenerator extends AbstractGenerator {

    public AllDifferentGenerator(final Problem problem) {
        super(problem);
    }

    @Override
    public boolean generate(final CSPOMConstraint constraint)
            throws FailedGenerationException {
        Preconditions.checkArgument(constraint instanceof GeneralConstraint,
                "Not supported");
        final List<Variable> solverVariables = Lists.transform(
                constraint.getScope(), CSPOM_TO_CSP4J);
        if (Iterables.any(solverVariables, NULL_DOMAIN)) {
            return false;
        }
        addConstraint(new AllDifferent(
                solverVariables.toArray(new Variable[solverVariables.size()])));
        return true;
    }

}

File: src/main/scala/cspfj/generator/constraint/AllDifferentGenerator.scala
package cspfj.generator.constraint;

import cspfj.{ Variable, Problem }
import cspom.constraint.{ GeneralConstraint, CSPOMConstraint }
import cspfj.constraint.semantic.{ AllDifferent2C, AllDifferentBC }

final class AllDifferentGenerator(problem: Problem) extends AbstractGenerator(problem) {
  override def generateGeneral(constraint: GeneralConstraint) = {
    val solverVariables = constraint.scope map cspom2cspfj
    if (solverVariables exists { _.dom == null }) {
      false
    } else {
      addConstraint(new AllDifferent2C(solverVariables: _*));
      addConstraint(new AllDifferentBC(solverVariables: _*));
      true;
    }
  }

}
